### ISSUE TITLE
WeaveNet started via weave-kube breaks multicast

### ISSUE THREAD

**Issue by bergtwvd** [weaveworks/weave#2689]
When I use Docker Compose on a cluster of docker hosts and weave overlay networking / multicast then I use the following declaration in the network section of the Docker compose file:

````
networks:
 training:
  driver: weave
  driver_opts:
   works.weave.multicast: "1"
````

I have set up Kubernetes with Weave Net with a couple of containers talking multicast.
How can I achieve on Kubernetes what I do on a Docker overlay network ?

(they  run fine, when deployed as a pod, but I want them to run on different nodes)

**Comment by bboreham** [comment]
I just tried this with the `weave-kube` deployment, and I saw that the Network Policy Controller is blocking multicast traffic.  You can check if you have the same problem by looking at the log files of the `weave-npc` container on each host.

We're just pondering how best to fix.  You can work around by disabling the firewall completely (delete the daemonset; edit the template to set env var `EXPECT_NPC` to `0` for weave-kube container, reboot and recreate the daemonset) or by adding a specific iptables rule on each host to allow multicast traffic.

**Comment by bergtwvd** [comment]
Just to confirm, I see the following message in the weave npc log on the node where the (multicast) pod runs:

``time="2016-12-12T23:00:00Z" level=warning msg="UDP connection from 10.42.0.4:20913 to 239.20.9.13:20913 blocked by Weave NPC."``

**Comment by bergtwvd** [comment]
Not being a kube expert yet, where / how can I edit the template to set EXPECT_NPC  ?

**Comment by bboreham** [comment]
@bergtwvd we give an example at https://www.weave.works/docs/net/latest/kube-addon/#configuration-options

**Comment by bboreham** [comment]
Some notes on the underlying problem:

`weave-npc` works via `iptables` rules saying (via ipset) which IP addresses can send to which other IP addresses.  At present no multicast addresses are in the sets, so all such traffic is blocked.

If we think of adding multicast to the set of allowed addresses, we have the problem of distinguishing which containers should receive it - the same multicast address may be used by multiple pods, but there is only one set of rules.

Trying to enumerate the rules that `weave-npc` should enforce, it should allow multicast traffic:
 1. from anywhere to pods in namespaces which are `DefaultAllow` (the default).
 1. to pods in namespaces which are `DefaultDeny`, if permitted by a rule which allows the port number
 1. to pods in namespaces which are `DefaultDeny`, from pod IP addresses specifically selected by a rule

It seems we need to add a rule in the host namespace to allow multicast, then add rules in each pod's namespace to enforce the above.  Which may change as policies change, or as pods appear and disappear under rule 3.

For fail-safeness we need to have the blocking rule in place before Kubernetes runs any code in the pod; this suggests an interlock is needed between the CNI plugin and the policy controller.

**Comment by bergtwvd** [comment]
I tried out your suggestion: delete daemonset, edit template, reboot, and recreate daemonset.
How ever I still see the message ``time="2016-12-13T21:33:37Z" level=warning msg="UDP connection from 10.42.0.7:20913 to 239.20.9.13:20913 blocked by Weave NPC."`` in the weave_npc log.

There is one observation with the env setting:

I downloaded the weave-daemonset.yaml file and added the environment variable:

````
...
      containers:
        - name: weave
          env:
            - name: EXPECT_NPC
              value: 0
...
````

This causes an error: 
``
[bergtwvd@kub-master weave]$ kubectl apply -f  weave-daemonset.yaml
unable to decode "weave-daemonset.yaml": [pos 469]: json: expect char '"' but got char '0'
``

So I put double quotes around the character.

Steps done, without effect:
- edited weave-daemonset.yaml
- delete daemonset via Kube UI
- reboot
- kubectl apply -f  weave-daemonset.yaml

**Comment by bergtwvd** [comment]
Tried it another way, with more success.

Rather than going via the UI, I did
``kubectl delete daemonset weave-net --namespace=kube-system``
from the master node.

This deleted not only the daemonset, but also the pods on each node.  Follows other steps, and now every container in every pod can char multicast, regardless of the node it is running on.

**Comment by bboreham** [comment]
Fixed (at least the key symptom of breaking multicast completely) by #2747

**Comment by skymir74** [comment]
In my company, dev team are making a PaaS solution working on the Kubernetest 
and they chose weave net as cni to support multicast traffic. 


Now my team have met with some difficulties as following:

1. Can drop(deny) multicast from other kubernetes service using kubernetes network policy? 

   My PaaS team want to transper a software solution, called M-CACHE, using multicast from old legacy system to kubernetes service. 
   M-CACHE uses multicasting manner to gather or distribute some information with its IGMP members. 
   
   My team have to operate two or more k8s services on a k8s cluster and each my k8s service contains a M-CACHE.
   It means that different k8s services contain MCACHEs which all have the same fixed IGMP configuration(1 multicast address).
   Unfortunately my team have to use a MCACHE without any change(including any configuration changes) due to proprietary reason. 

   at this point, my team expected that 
   with k8s network policy on each services or namespaces,
   multicast traffics from different k8s services may be blocked just as other tcp/ip traffics. 
   
   But I check that no network policy can block any multicast traffic from other pod/namespace/service until now.  
   Do you offer proper way to block multicast traffic from other k8s pod, namespace, or service? 
   Or, if not supported yet, do you have any plan to support any network policy about multicast traffic?
   
2. Can kubernetes with weavenet support multitenant service?

   My team want to support the same k8s service with different subnet address. 
   Can kubernetes with weavenet assign different subnet address on each k8s service?
   for example, k8s service A with 11.1.1.0/24 and k8s service B with 11.1.2.0/24. 
   
   
   At this present, my team makes k8s system without cloud service, such as AWS, GCE or openstack. 
  
   I'm not sure whether kubernetes with weavenet can support multitenant service or not 
   Is there any other ways kubernetes with weavenet can support multitenency without cloud service?

**Comment by bboreham** [comment]
Please don’t comment on an old, closed, issue. Open your own issues so the thread of conversation is clear. 

1. I think you are saying you want to allow some multicast conversations but not others, where the IP addresses used are always the same. To do this would need a very different implementation to the current Weave-npc.

2. Not going to answer here because it’s not even vaguely related to this issue.  Open a new issue.
